src/librbd/ImageWatcher.cc

```diff
@@ -748,6 +748,16 @@ void ImageWatcher<I>::schedule_async_request_timed_out(const AsyncRequestId &id)
 
 template <typename I>
 void ImageWatcher<I>::async_request_timed_out(const AsyncRequestId &id) {
+
+  // to avoid a window where the request is not yet on m_async_requests
+  auto expiration_time = ceph_clock_now() + 60;
+  auto it = m_async_requests.find(id);
+  while (it != m_async_requests.end() && expiration_time < ceph_clock_now()) {
+    if (it->second.second != nullptr) {
```

Comment:
Ilya, as you know I'm not able to reproduce this issue straight.

I thought you suggested this solution:

>> Also, Have we thought about any possible solution yet?

> One possible solution might be to massage ImageWatcher::async_request_timed_out() and the surrounding code so that the short window where the request is not on m_async_requests (and therefore can't be found, leading to AsyncCompletePayload message getting lost) is avoided. Another might be actually resend AsyncCompletePayload message, if feasible.

It looks like I misinterpreted something, let's talk on the call.

Reply:
Ah! I get you now, sent the wrong fix, I mean to check for AsyncRequest is empty. Not it->second.second :-(

Reply:
> Ilya, as you know I'm not able to reproduce this issue straight.
> 
> I thought you suggested this solution:

Any proposed solution is just that -- a proposal -- until it's verified to work.  Race conditions like this are in fact tough to reproduce but as I mentioned in the tracker it may be doable with some delays injected in the right spots.  In the absence of a reproducer, a crystal clear explanation of both the race and the fix must be provided -- otherwise there is no just way to judge it.

> It looks like I misinterpreted something, let's talk on the call.

Sure!